### PR TITLE
Add  verifyHostStateForDrive

### PR DESCRIPTION
**Pull request by sinhaashish**

### CODE REVIEW

File: pkg/drive/utils.go [hunk 1]
```diff
@@ -59,5 +74,67 @@ func getFSUUIDFromDrive(drive *directcsi.DirectCSIDrive) string {
...
+
+	devName, err := sys.GetDeviceName(uint32(drive.Status.MajorNumber), uint32(drive.Status.MinorNumber))
+	if err != nil {
+		return err
+	}
+	if filepath.Base(drive.Status.Path) != devName {
+		return fmt.Errorf("path mismatch. Expected %s got %s", filepath.Base(drive.Status.Path), devName)
+	}
+
+	if !utils.IsCondition(drive.Status.Conditions,
```
Comment (Praveenrajmani) [inline pkg/drive/utils.go]: we should error out if the ready condition is false. no need to check for specific reason. refer IsConditionStatus function..

File: pkg/drive/utils.go [hunk 2]
```diff
@@ -59,5 +74,67 @@ func getFSUUIDFromDrive(drive *directcsi.DirectCSIDrive) string {
...
+		DMUUID:       runUdevData.DMUUID,
+		MDUUID:       runUdevData.MDUUID,
+		PTUUID:       runUdevData.PTUUID,
+		PTType:       runUdevData.PTType,
+		PartUUID:     runUdevData.PartUUID,
+		UeventFSUUID: runUdevData.UeventFSUUID,
+		FSType:       runUdevData.FSType,
+	}
+
+	if !uevent.ValidateDevice(device, []*directcsi.DirectCSIDrive{drive}) {
```
Comment (Praveenrajmani) [inline pkg/drive/utils.go]: we need to move ValidateDevice to sys pkg as discussed

File: pkg/drive/utils.go [hunk 3]
```diff
@@ -59,5 +74,67 @@ func getFSUUIDFromDrive(drive *directcsi.DirectCSIDrive) string {
...
+
+	if drive.Status.DriveStatus == directcsi.DriveStatusInUse ||
+		drive.Status.DriveStatus == directcsi.DriveStatusReady {
+		mounted, err := mount.IsMounted(drive.Name)
+		if err != nil {
+			return err
+		}
+		if mounted {
+			if !node.ValidDirectPVMountOpts(drive.Status.MountOptions) {
+				return errInvalidMountOptions
```
Comment (Praveenrajmani) [inline pkg/drive/utils.go:130]: ValidDirectPVMountOpts can be moved to mnt pkg and used commonly from there

File: pkg/sys/sys_utils_test.go [hunk 4]
```diff
@@ -19,6 +19,7 @@ package sys
 import (
 	"reflect"
 	"testing"
+	//	"github.com/minio/directpv/pkg/sys"
```
Comment (Praveenrajmani) [inline pkg/sys/sys_utils_test.go]: this comment can be removed

File: pkg/drive/utils.go (hunk 2 above)
Comment (sinhaashish) [reply pkg/drive/utils.go]: it was erroring because of circular dependency

File: pkg/drive/utils.go (hunk 3 above)
Comment (sinhaashish) [reply pkg/drive/utils.go:130]: done

File: pkg/drive/utils.go [hunk 5]
```diff
@@ -17,9 +17,23 @@
...
+	"github.com/minio/directpv/pkg/mount"
 	"github.com/minio/directpv/pkg/sys"
+	"github.com/minio/directpv/pkg/uevent"
+	"github.com/minio/directpv/pkg/utils"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	errInvalidDevPath     = errors.New("devpath not found in the event")
+	errDriveValueMismatch = errors.New("Drive value mismatch")
```
Comment (Praveenrajmani) [inline pkg/drive/utils.go]:
```suggestion
	errDriveValueMismatch = errors.New("drive value mismatch")
```

File: pkg/drive/utils.go [hunk 6]
```diff
@@ -59,5 +73,67 @@ func getFSUUIDFromDrive(drive *directcsi.DirectCSIDrive) string {
...
+		return err
+	}
+	if filepath.Base(drive.Status.Path) != devName {
+		return fmt.Errorf("path mismatch. Expected %s got %s", filepath.Base(drive.Status.Path), devName)
+	}
+
+	if utils.IsConditionStatus(drive.Status.Conditions,
+		string(directcsi.DirectCSIDriveConditionReady),
+		metav1.ConditionFalse) {
+		return fmt.Errorf("drive is lost = %v", drive.Status.Conditions)
```
Comment (Praveenrajmani) [inline pkg/drive/utils.go]:
```suggestion
		return fmt.Errorf("drive %s is not ready", drive.Name)
```

File: pkg/uevent/utils.go [hunk 7]
```diff
@@ -40,7 +40,7 @@ func getRootBlockPath(devName string) string {
 	}
 }
 
-func validateDevice(device *sys.Device, filteredDrives []*directcsi.DirectCSIDrive) bool {
+func ValidateDevice(device *sys.Device, filteredDrives []*directcsi.DirectCSIDrive) bool {
```
Comment (Praveenrajmani) [inline pkg/uevent/utils.go]: we can export a single validate function which accepts a device and a drive

Like, 

```
func ValidateDriveStates(device *sys,Device, directCSIDrive *directcsi.DirectCSIDrive) bool {
        if directCSIDrive.Status.Path != device.DevPath() {
		return false
	}
	if directCSIDrive.Status.MajorNumber != uint32(device.Major) {
		return false
	}
	if directCSIDrive.Status.MinorNumber != uint32(device.Minor) {
		return false
	}
	if directCSIDrive.Status.Virtual != device.Virtual {
		return false
	}
	if directCSIDrive.Status.PartitionNum != device.Partition {
		return false
	}
	if directCSIDrive.Status.WWID != device.WWID {
		return false
	}
	if directCSIDrive.Status.ModelNumber != device.Model {
		return false
	}
	if directCSIDrive.Status.UeventSerial != device.UeventSerial {
		return false
	}
	if directCSIDrive.Status.Vendor != device.Vendor {
		return false
	}
	if directCSIDrive.Status.DMName != device.DMName {
		return false
	}
	if directCSIDrive.Status.DMUUID != device.DMUUID {
		return false
	}
	if directCSIDrive.Status.MDUUID != device.MDUUID {
		return false
	}
	if directCSIDrive.Status.PartTableUUID != device.PTUUID {
		return false
	}
	if directCSIDrive.Status.PartTableType != device.PTType {
		return false
	}
	if directCSIDrive.Status.PartitionUUID != device.PartUUID {
		return false
	}
	if directCSIDrive.Status.Filesystem != device.FSType {
		return false
	}
	if directCSIDrive.Status.UeventFSUUID != device.UeventFSUUID {
		return false
	}

	return true
}
```

and use this function here in L47. 

So, that we don't need to pass a list of drive as an input from the other places.

File: pkg/drive/utils.go (hunk 2 above)
Comment (Praveenrajmani) [reply pkg/drive/utils.go]: both `pkg drive` and `pkg uevent` already imports and works with sys pkg. so, there won't be cyclic dependency

Comment (sinhaashish) [reply pkg/drive/utils.go]: yes but the signature of the function is `func ValidateDevice(device *Device, directCSIDrive *directcsi.DirectCSIDrive) bool {`  and `directcsi` - > `"github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta3"` 

This leads to cyclic dependency 
```
package github.com/minio/directpv/cmd/directpv
	imports github.com/minio/directpv/pkg/client
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1
	imports github.com/minio/directpv/pkg/matcher
	imports github.com/minio/directpv/pkg/sys
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta3
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta2
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1: import cycle not allowed
package github.com/minio/directpv/cmd/directpv
	imports github.com/minio/directpv/pkg/client
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1
	imports github.com/minio/directpv/pkg/matcher
	imports github.com/minio/directpv/pkg/sys
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta3
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta2
	imports github.com/minio/directpv/pkg/matcher: import cycle not allowed
package github.com/minio/directpv/cmd/directpv
	imports github.com/minio/directpv/pkg/client
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1
	imports github.com/minio/directpv/pkg/matcher
	imports github.com/minio/directpv/pkg/sys
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta3
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta2
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1: import cycle not allowed
package github.com/minio/directpv/cmd/directpv
	imports github.com/minio/directpv/pkg/client
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta1
	imports github.com/minio/directpv/pkg/matcher
	imports github.com/minio/directpv/pkg/sys
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta3
	imports github.com/minio/directpv/pkg/apis/direct.csi.min.io/v1beta2
	imports github.com/minio/directpv/pkg/matcher: import cycle not allowed

```

**Review by Praveenrajmani — APPROVED** [review]

**Comment by Praveenrajmani** [comment]
can you also merge this PR https://github.com/sinhaashish/directpv/pull/2 to your branch @sinhaashish 

i have added some more fixes which were found during testing.

**Comment by Praveenrajmani** [comment]
this PR can be closed as the changes are merged in https://github.com/minio/directpv/pull/554

**Pull request closed without merging**